src/bvh/bvh.rs

```diff
@@ -0,0 +1,444 @@
+//! This module defines a [`BVH`] building procedure as well as a [`BVH`] flattening procedure
+//! so that the recursive structure can be easily used in compute shaders.
+//!
+//! [`BVH`]: struct.BVH.html
+//!
+
+use EPSILON;
+use aabb::{AABB, Bounded};
+use bounding_hierarchy::{BoundingHierarchy, BHShape};
+use ray::Ray;
+use std::f32;
+use std::iter::repeat;
+
+/// Enum which describes the union type of a node in a [`BVH`].
+/// This structure does not allow for storing a root node's [`AABB`]. Therefore rays
+/// which do not hit the root [`AABB`] perform two [`AABB`] tests (left/right) instead of one.
+/// On the other hand this structure decreases the total number of indirections when traversing
+/// the BVH. Only those nodes are accessed, which are definetely hit.
+///
+/// [`AABB`]: ../aabb/struct.AABB.html
+/// [`BVH`]: struct.BVH.html
+///
+#[derive(Debug, Copy, Clone)]
+pub enum BVHNode {
+    /// Leaf node.
+    Leaf {
+        /// The node's parent.
+        parent: usize,
+
+        /// The node's depth
+        depth: u32,
```

Comment:
I guess i could as well use usize for the depth, but I don't see why.
Except maybe that a very degenerated tree can have a depth as high as the highest index.

To me, usize kinda has this semantic aftertaste of being used for pointer-like things, like slice accessing. That's why i went for u32 instead, which feels more like a pure, generic number type. But maybe that's just me.